nvmeadm/src/nvmf_subsystem.rs

```diff
@@ -23,26 +23,114 @@ pub struct Subsystem {
+    Unknown { key: String, value: String },
+}
+impl From<(&str, &str)> for SubsystemAddrToken {
+    fn from((key, value): (&str, &str)) -> Self {
+        let value = value.to_string();
+        match key {
+            TR_ADDR => Self::TrAddr { traddr: value },
+            TR_SVC_ID => Self::TrSvcId { trsvcid: value },
+            SRC_ADDR => Self::SrcAddr { src_addr: value },
+            unknown => Self::Unknown {
+                key: unknown.to_string(),
+                value,
+            },
+        }
+    }
+}
+
+/// A slightly parsed version of `SubsystemAddr` containing the address tokens as variables
+/// making it easier to retrieve them.
+#[derive(Debug, Default, Clone)]
+pub struct SubsystemAddrExt {
+    tr_addr: Option<String>,
+    tr_svc_id: Option<String>,
+    src_addr: Option<String>,
+    unknowns: Vec<SubsystemAddrToken>,
+}
+
+impl SubsystemAddrExt {
+    /// Check if the subsystem address contains target port.
+    pub fn match_host_port(&self, host: &str, port: &str) -> bool {
+        self.tr_addr.as_deref() == Some(host) && self.tr_svc_id.as_deref() == Some(port)
+    }
+}
+
+impl From<SubsystemAddr> for SubsystemAddrExt {
+    fn from(value: SubsystemAddr) -> Self {
+        Self::from(value.0.as_str())
+    }
+}
+// This would be nice to do with serde, but for now let's not include serde in this crate
+impl From<&str> for SubsystemAddrExt {
+    fn from(value: &str) -> Self {
+        let raw_addr = value.split(',');
+        raw_addr.fold(SubsystemAddrExt::default(), |mut acc, addr| {
+            if let Some(pair) = addr.split_once('=') {
+                match SubsystemAddrToken::from(pair) {
+                    SubsystemAddrToken::TrAddr { traddr } => {
+                        acc.tr_addr = Some(traddr);
+                    }
+                    SubsystemAddrToken::TrSvcId { trsvcid } => {
+                        acc.tr_svc_id = Some(trsvcid);
+                    }
+                    SubsystemAddrToken::SrcAddr { src_addr } => {
+                        acc.src_addr = Some(src_addr);
+                    }
+                    addr @ SubsystemAddrToken::Unknown { .. } => acc.unknowns.push(addr),
+                }
+            }
+            acc
+        })
+    }
+}
+
+/// Wrapper over a raw subsystem address string.
+/// The address contains a comma-separated list of `SubsystemAddrToken`, example:
+/// traddr=10.1.0.2,trsvcid=8420.
+#[derive(Clone, Debug)]
 pub struct SubsystemAddr(String);
 
 impl SubsystemAddr {
-    /// New SubsystemAddr.
-    pub fn new(host: String, port: u16) -> SubsystemAddr {
-        SubsystemAddr(format!("traddr={host},trsvcid={port}"))
+    /// Check if the subsystem address contains target port.
+    pub fn match_host_port(&self, host: &str, port: &str) -> bool {
+        SubsystemAddrExt::from(self.as_str()).match_host_port(host, port)
+    }
+    /// Collect all address variants into a key-val map.
+    pub fn collect(&self) -> HashMap<&str, &str> {
```

Comment:
Is the collect function used?

Reply:
No, left it as part of pub in case we need to inspect the key-map